src/coreclr/jit/helperexpansion.cpp

```diff
@@ -2268,7 +2279,18 @@ bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt,
     }
     else
     {
-        GenTree* fallbackTree = gtNewTempStore(tmpNum, call);
+        GenTreeCall* helperCall = call;
+        if (typeCheckFailedAction == TypeCheckFailedAction::CallHelper_Specialized)
+        {
+            // Re-create the call with a specialized helper
+            // For JIT we probably can just change the gtCallMethHnd in the existing call, but let's unify with AOT
+            helperCall = gtNewHelperCallNode(CORINFO_HELP_CHKCASTCLASS_SPECIAL, call->TypeGet(),
+                                             helperCall->gtArgs.GetUserArgByIndex(0)->GetNode(),
+                                             helperCall->gtArgs.GetUserArgByIndex(1)->GetNode());
+            fgMorphCall(helperCall);
+            gtSetEvalOrder(helperCall);
```

Comment:
Just curious -- how is NAOT and JIT different here?

Reply:
AOT needs you to set the entry point. This could be done by mutating the node, theoretically.

Reply:
Hmm, but what is setting the entry point here?

Reply:
Good point - nothing, so that is not the reason then.

Reply:
Ah, it's not needed at all - `CORINFO_HELP_CHKCASTCLASS` is not used for AOT anyway so we're only improving JIT here (copies importer's behavior)

Reply:
Ah, looks like it's needed if I were changing entrypoint for CORINFO_HELP_READYTORUN_CHKCAST